[Model] Add metapath2vec package

LuckyLiuM opened the pull request:
## Description
<!-- Brief description. Refer to the related issues if existed.
It'll be great if relevant reviewers can be assigned as well.-->
main2.py is the package I write and integrate based on examples
cac_test.py is the training process using my model in main2
test.py is the test file

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

## Changes
<!-- You could use following template
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)
-->

mufeili commented:
cc @zhjwy9343 as you requested this feature before.

mufeili commented on examples/pytorch/metapath2vec/cac_test.py:
```diff
@@ -0,0 +1,184 @@
+		return hg,self.id_conf,self.id_author
+
+def main():
+	#参数设置
```
Could you change the comments to English?

mufeili commented on examples/pytorch/metapath2vec/main2.py:
```diff
@@ -0,0 +1,212 @@
+import torch
```
You need to move this module to `dgl/python/dgl/nn/pytorch/`. You can create a file `network_emb.py` there.

mufeili commented:
Could you directly use this built-in `Metapath2vec` module in the old example and replace the related code?

mufeili commented:
Could you add unit tests in https://github.com/dmlc/dgl/blob/master/tests/pytorch/test_nn.py to test `Metapath2vec`? You can generate some synthetic data if needed.

LuckyLiuM commented:
>

LuckyLiuM commented:
> Could you directly use this built-in `Metapath2vec` module in the old example and replace the related code?

I have replaced the old examples and deleted the unnecessary files. There are many difference between the old example and the new one. I delete the dataset class and default download path because the new model 's input is a DGLgraph rather than metapath data in the old one.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        ##start random walk
+        for idx in tqdm.trange(self.g.number_of_nodes(nodespath[0])):
+            traces = random_walk(g=self.g, nodes=[idx, ] * self.node_repeat, metapath=self.metapath)
+            for tr in traces[0].numpy():
```
```suggestion
            for tr in traces[0].cpu().numpy():
```

mufeili commented:
You need to update the `README.md` under `dgl/examples/pytorch/metapath2vec/` so that others can run your code.

mufeili commented:
We typically use 4 white spaces for indent while you are using a mixture of 4 and 8.

mufeili commented on tests/pytorch/test_nn.py:
```diff
@@ -1621,4 +1621,27 @@ def test_dgn_conv(in_size, out_size, aggregators, scalers, delta,
+
+
+@parametrize_idtype
+@pytest.mark.parametrize('meta_path', [['uc','cu'],['uc','cp','pc','cu'],['cp','pc','cp','pc']])
```
I think it suffices to try two options here.

mufeili commented on tests/pytorch/test_nn.py:
```diff
@@ -1621,4 +1621,27 @@ def test_dgn_conv(in_size, out_size, aggregators, scalers, delta,
+
+@parametrize_idtype
+@pytest.mark.parametrize('meta_path', [['uc','cu'],['uc','cp','pc','cu'],['cp','pc','cp','pc']])
+@pytest.mark.parametrize('window_size', [1,3])
```
```suggestion
@pytest.mark.parametrize('window_size', [1, 3])
```

mufeili commented on tests/pytorch/test_nn.py:
```diff
@@ -1621,4 +1621,27 @@ def test_dgn_conv(in_size, out_size, aggregators, scalers, delta,
+@parametrize_idtype
+@pytest.mark.parametrize('meta_path', [['uc','cu'],['uc','cp','pc','cu'],['cp','pc','cp','pc']])
+@pytest.mark.parametrize('window_size', [1,3])
+@pytest.mark.parametrize('negative_samples', [1,3])
```
```suggestion
@pytest.mark.parametrize('negative_samples', [1, 3])
```

mufeili commented on tests/pytorch/test_nn.py:
```diff
@@ -1621,4 +1621,27 @@ def test_dgn_conv(in_size, out_size, aggregators, scalers, delta,
+@pytest.mark.parametrize('meta_path', [['uc','cu'],['uc','cp','pc','cu'],['cp','pc','cp','pc']])
+@pytest.mark.parametrize('window_size', [1,3])
+@pytest.mark.parametrize('negative_samples', [1,3])
+@pytest.mark.parametrize('min_count', [0,1])
```
```suggestion
@pytest.mark.parametrize('min_count', [0, 1])
```

mufeili commented on tests/pytorch/test_nn.py:
```diff
@@ -1621,4 +1621,27 @@ def test_dgn_conv(in_size, out_size, aggregators, scalers, delta,
+@pytest.mark.parametrize('window_size', [1,3])
+@pytest.mark.parametrize('negative_samples', [1,3])
+@pytest.mark.parametrize('min_count', [0,1])
+@pytest.mark.parametrize('emb_dim', [1,  100])
```
```suggestion
@pytest.mark.parametrize('emb_dim', [1, 100])
```

mufeili commented on tests/pytorch/test_nn.py:
```diff
@@ -1621,4 +1621,27 @@ def test_dgn_conv(in_size, out_size, aggregators, scalers, delta,
+@pytest.mark.parametrize('negative_samples', [1,3])
+@pytest.mark.parametrize('min_count', [0,1])
+@pytest.mark.parametrize('emb_dim', [1,  100])
+@pytest.mark.parametrize('node_repeat', [1,  3])
```
```suggestion
@pytest.mark.parametrize('node_repeat', [1, 3])
```

mufeili commented on tests/pytorch/test_nn.py:
```diff
@@ -1621,4 +1621,27 @@ def test_dgn_conv(in_size, out_size, aggregators, scalers, delta,
+@pytest.mark.parametrize('min_count', [0,1])
+@pytest.mark.parametrize('emb_dim', [1,  100])
+@pytest.mark.parametrize('node_repeat', [1,  3])
+def test_Metapath2vec(emb_dim,meta_path,window_size,negative_samples,min_count,node_repeat,idtype):
```
```suggestion
def test_Metapath2vec(emb_dim, meta_path, window_size, negative_samples, min_count, node_repeat, idtype):
```

mufeili replied:
It may not be necessary to test different options for all these arguments. This yields a large product space.

mufeili commented on tests/pytorch/test_nn.py:
```diff
@@ -1621,4 +1621,27 @@ def test_dgn_conv(in_size, out_size, aggregators, scalers, delta,
+def test_Metapath2vec(emb_dim,meta_path,window_size,negative_samples,min_count,node_repeat,idtype):
+
+    g = dgl.heterograph({
+        ('user', 'uc', 'company'): ([0, 0, 2, 1,3], [1, 2, 1, 3,0]),
```
```suggestion
        ('user', 'uc', 'company'): ([0, 0, 2, 1, 3], [1, 2, 1, 3, 0]),
```

mufeili commented on tests/pytorch/test_nn.py:
```diff
@@ -1621,4 +1621,27 @@ def test_dgn_conv(in_size, out_size, aggregators, scalers, delta,
+
+    g = dgl.heterograph({
+        ('user', 'uc', 'company'): ([0, 0, 2, 1,3], [1, 2, 1, 3,0]),
+        ('company', 'cp', 'product'): ([0, 0, 0, 1, 2,3], [0, 2, 3, 0, 2,1]),
```
```suggestion
        ('company', 'cp', 'product'): ([0, 0, 0, 1, 2, 3], [0, 2, 3, 0, 2, 1]),
```

mufeili commented on tests/pytorch/test_nn.py:
```diff
@@ -1621,4 +1621,27 @@ def test_dgn_conv(in_size, out_size, aggregators, scalers, delta,
+    g = dgl.heterograph({
+        ('user', 'uc', 'company'): ([0, 0, 2, 1,3], [1, 2, 1, 3,0]),
+        ('company', 'cp', 'product'): ([0, 0, 0, 1, 2,3], [0, 2, 3, 0, 2,1]),
+        ('company', 'cu', 'user'): ([1, 2, 1, 3,0], [0, 0, 2, 1,3]),
```
```suggestion
        ('company', 'cu', 'user'): ([1, 2, 1, 3, 0], [0, 0, 2, 1, 3]),
```

mufeili commented on tests/pytorch/test_nn.py:
```diff
@@ -1621,4 +1621,27 @@ def test_dgn_conv(in_size, out_size, aggregators, scalers, delta,
+        ('company', 'cp', 'product'): ([0, 0, 0, 1, 2,3], [0, 2, 3, 0, 2,1]),
+        ('company', 'cu', 'user'): ([1, 2, 1, 3,0], [0, 0, 2, 1,3]),
+        ('product', 'pc', 'company'): ([0, 2, 3, 0, 2,1], [0, 0, 0, 1, 2,3])
+    },idtype=idtype, device=F.ctx())
```
```suggestion
    }, idtype=idtype, device=F.ctx())
```

mufeili commented on tests/pytorch/test_nn.py:
```diff
@@ -1621,4 +1621,27 @@ def test_dgn_conv(in_size, out_size, aggregators, scalers, delta,
+        ('company', 'cu', 'user'): ([1, 2, 1, 3,0], [0, 0, 2, 1,3]),
+        ('product', 'pc', 'company'): ([0, 2, 3, 0, 2,1], [0, 0, 0, 1, 2,3])
+    },idtype=idtype, device=F.ctx())
+    model = nn.Metapath2vec(g,emb_dim,meta_path,window_size,min_count,negative_samples,node_repeat)
```
```suggestion
    model = nn.Metapath2vec(g, emb_dim, meta_path, window_size, min_count, negative_samples, node_repeat)
```

mufeili commented on tests/pytorch/test_nn.py:
```diff
@@ -1621,4 +1621,27 @@ def test_dgn_conv(in_size, out_size, aggregators, scalers, delta,
+    },idtype=idtype, device=F.ctx())
+    model = nn.Metapath2vec(g,emb_dim,meta_path,window_size,min_count,negative_samples,node_repeat)
+    model = model.to(F.ctx())
+    embeds=model.u_embeddings.weight
```
```suggestion
    embeds = model.u_embeddings.weight
```

mufeili commented on tests/pytorch/test_nn.py:
```diff
@@ -1621,4 +1621,27 @@ def test_dgn_conv(in_size, out_size, aggregators, scalers, delta,
+    model = nn.Metapath2vec(g,emb_dim,meta_path,window_size,min_count,negative_samples,node_repeat)
+    model = model.to(F.ctx())
+    embeds=model.u_embeddings.weight
+    emb_size=model.word_count
```
```suggestion
    emb_size = model.word_count
```

mufeili commented on tests/pytorch/test_nn.py:
```diff
@@ -1621,4 +1621,27 @@ def test_dgn_conv(in_size, out_size, aggregators, scalers, delta,
+    model = model.to(F.ctx())
+    embeds=model.u_embeddings.weight
+    emb_size=model.word_count
+    assert embeds.shape==(emb_size,emb_dim)
```
```suggestion
    assert embeds.shape == (emb_size, emb_dim)
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+import tqdm
+from ...sampling import random_walk
+
+
```
```suggestion
__all__ = ['Metapath2vec']
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+
+
+
+class Metapath2vec(nn.Module):
```
```suggestion
class MetaPath2Vec(nn.Module):
```

We use the camel naming convention.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+
+class Metapath2vec(nn.Module):
+    """
+    Metapath2vec, a scalable representation learning model for heterogeneous networks
```
```suggestion
    metapath2vec module from `metapath2vec: Scalable Representation Learning for
Heterogeneous Networks <https://dl.acm.org/doi/pdf/10.1145/3097983.3098036>`__
```

to be consistent with existing modules

Split it into two lines if it is too long.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+class Metapath2vec(nn.Module):
+    """
+    Metapath2vec, a scalable representation learning model for heterogeneous networks
+    from KDD’17, August <https://dl.acm.org/doi/pdf/10.1145/3097983.3098036>
```
Remove this line.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+    """
+    Metapath2vec, a scalable representation learning model for heterogeneous networks
+    from KDD’17, August <https://dl.acm.org/doi/pdf/10.1145/3097983.3098036>
+    Parameters
```
Add an empty line above.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+    from KDD’17, August <https://dl.acm.org/doi/pdf/10.1145/3097983.3098036>
+    Parameters
+    ----------
+    g :DGLHeteroGraph
```
```suggestion
    g : DGLGraph
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+    ----------
+    g :DGLHeteroGraph
+        Graph data for generating embedding
+    emb_dimension :int
```
```suggestion
    emb_dimension : int
```

mufeili replied:
Personally I prefer calling it `emb_dim`, which is shorter.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        Graph data for generating embedding
+    emb_dimension :int
+        Size of each embedding vector
+    metapath :List[str]
```
```suggestion
    metapath : List[str]
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        the value is its corresponding name.(default: None)
+    sparse  :int , optional
+        If set to True, gradients w.r.t. to the weight matrix will be sparse. (default: True)
+    """
```
Add an `Examples` section.

Reference:
- https://docs.dgl.ai/generated/dgl.metapath_reachable_graph.html?highlight=metapath#dgl.metapath_reachable_graph
- https://docs.dgl.ai/generated/dgl.nn.pytorch.explain.GNNExplainer.html#dgl.nn.pytorch.explain.GNNExplainer

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        The metapath is described as a list of strings for relation edges between different
+        types of nodes. It is a  sequence of edges which denotes transition  for different
+        types of nodes at each step. Note that the node type at the start and end of the
+        metapath are commonly the same.
```
I'll probably reword the description of `metapath` as follows.

```
A sequence of edge types in the form of string. It defines a new edge type by composing multiple edge types in order. Note that the start node type and the end one must be the same.
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        types of nodes. It is a  sequence of edges which denotes transition  for different
+        types of nodes at each step. Note that the node type at the start and end of the
+        metapath are commonly the same.
+    window_size :int
```
```suggestion
    window_size : int
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        types of nodes at each step. Note that the node type at the start and end of the
+        metapath are commonly the same.
+    window_size :int
+        The context size which is considered for positive samples.
```
Is "context" and "window" equivalent in this case? If so, why don't we call it `context_size` instead?

mufeili replied:
Perhaps you need to explain what are positive and negative samples at the beginning and how they affect the training.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        metapath are commonly the same.
+    window_size :int
+        The context size which is considered for positive samples.
+    min_count :int , optional
```
```suggestion
    min_count : int, optional
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        The context size which is considered for positive samples.
+    min_count :int , optional
+        The number of node counts that occur less than  min_count  value in all meta-paths
+        will be discarded before random walks. (default: 0)
```
Do you mean the sentence below?

```
The nodes that appear less than :attr:`min_count` times in all meta-path instances will be discarded.
```

What does it mean for them to be discarded before random walks? Does it mean that we will not obtain their embeddings at the end?

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+    min_count :int , optional
+        The number of node counts that occur less than  min_count  value in all meta-paths
+        will be discarded before random walks. (default: 0)
+    negative_samples_size :int , optional
```
How about calling it `num_negative` or `negative_size`?

mufeili replied:
```suggestion
    negative_samples_size : int, optional
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        The number of node counts that occur less than  min_count  value in all meta-paths
+        will be discarded before random walks. (default: 0)
+    negative_samples_size :int , optional
+        The number of negative samples need to be extracted for each positive sample.(default: 5)
```
```suggestion
        The number of negative samples to use for each positive sample. Default: 5
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        will be discarded before random walks. (default: 0)
+    negative_samples_size :int , optional
+        The number of negative samples need to be extracted for each positive sample.(default: 5)
+    node_repeat  :int , optional
```
How about calling it num_random_walk or random_walk_size?

mufeili replied:
```suggestion
    node_repeat : int, optional
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+    negative_samples_size :int , optional
+        The number of negative samples need to be extracted for each positive sample.(default: 5)
+    node_repeat  :int , optional
+        The number of random walks to sample for each start node$V_1$. (default: 1)
```
```suggestion
        The number of random walks to sample for each start node. Default: 1
```

You did not mention what is `$V_1$` at all. Also many python projects, including DGL, uses sphinx for building documentation. With sphinx, you need to use

```
:math:`V_1`
```

rather than

```
$V_1$
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        The number of negative samples need to be extracted for each positive sample.(default: 5)
+    node_repeat  :int , optional
+        The number of random walks to sample for each start node$V_1$. (default: 1)
+    nid2word  :Dict[ str, Dict[ int , str ] ], optional
```
When do we want to use this?

mufeili replied:
```suggestion
    nid2word : Dict[str, Dict[int, str]], optional
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+    node_repeat  :int , optional
+        The number of random walks to sample for each start node$V_1$. (default: 1)
+    nid2word  :Dict[ str, Dict[ int , str ] ], optional
+        If set, we can use model.id2word  to get the dict where the key is embedding id and
```
Should "model.id2word" be "model.nid2word" instead?

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        The number of random walks to sample for each start node$V_1$. (default: 1)
+    nid2word  :Dict[ str, Dict[ int , str ] ], optional
+        If set, we can use model.id2word  to get the dict where the key is embedding id and
+        the value is its corresponding name.(default: None)
```
Is this only useful when each node/entity corresponds to a string?

mufeili replied:
I will probably reword this description as follows.

:attr:`nid2word[node_type][node_id]` gives the string corresponding to the node of id :attr:`node_id` and type :attr:`node_type`.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+    nid2word  :Dict[ str, Dict[ int , str ] ], optional
+        If set, we can use model.id2word  to get the dict where the key is embedding id and
+        the value is its corresponding name.(default: None)
+    sparse  :int , optional
```
We need to test this argument in the unit test. Perhaps you can have some basic backward test.

mufeili replied:
```suggestion
    sparse : bool, optional
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        If set, we can use model.id2word  to get the dict where the key is embedding id and
+        the value is its corresponding name.(default: None)
+    sparse  :int , optional
+        If set to True, gradients w.r.t. to the weight matrix will be sparse. (default: True)
```
"weight matrix" -> "learnable weights"

mufeili replied:
```suggestion
        If set to True, gradients with respect to the learnable weights will be sparse. Default: True
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+    sparse  :int , optional
+        If set to True, gradients w.r.t. to the weight matrix will be sparse. (default: True)
+    """
+
```
Remove the empty line above

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+                 nid2word=None,
+                 sparse=True
+                 ):
+        super(Metapath2vec, self).__init__()
```
```suggestion
        super().__init__()
```

should be enough now

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        self.nid2word = nid2word
+        self.sparse = sparse
+        self.node_repeat = node_repeat
+        self.token_count = 0
```
It seems that `self.token_count` is only used in `_extract`, which always set `self.token_count = 0`. I think you can drop this line.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        self.negatives = []
+        self.discards = dict()
+        self.negpos = 0
+        self.NEGATIVE_TABLE_SIZE = 1e8
```
Perhaps expose this in `__init__`.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        np.random.shuffle(self.negatives)
+        self.sampling_prob = ratio
+
+    def initParameters(self):
```
Does this serve the same purpose as `reset_parameters` [here](https://docs.dgl.ai/generated/dgl.nn.pytorch.conv.GraphConv.html#dgl.nn.pytorch.conv.GraphConv)? If so, I suggest renaming this function for consistency.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+                    assert v < self.word_count
+                    if i == j:
+                        continue
+                    pair_catch.append((u, v, self._getNegatives(self.negative_samples_size)))
```
This is the only place where `_getNegatives` is used and the size is always `self.negative_samples_size`. Then why don't we drop the size argument of `_getNegatives` and let it directly access `self.negative_samples_size`?

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        self._initTableNegatives()
+
+    def _extract(self):
+
```
Remove the extra empty line here.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+
+    def _extract(self):
+
+        assert len(self.metapath) + 1 >= self.window_size
```
Perhaps add a comment that explains why this needs to be satisfied.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+
+        ##build whole vocab for all nodes
+        self.token_count = 0
+        self.type2id = dict()
```
I prefer renaming `type2id` to `local_to_global_id`. Essentially you are mapping type-specific node IDs to global node IDs.

mufeili replied:
You probably don't need to have `type2id` for all node types, only those involved in `metapath`.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+                self.type2id[nodetype] = dict()
+                for idx in range(self.g.num_nodes(nodetype)):
+                    self.type2id[nodetype][idx] = wid
+                    wid += 1
```
This can be inefficient. I might need to revisit it and see if we can drop or optimize it.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        ## convert the edge metapath to the node metapath
+        edge2des = {}
+        edge2src = {}
+        for comb in self.g.canonical_etypes:
```
```suggestion
        for src_type, etype, dst_type in self.g.canonical_etypes:
```
This might make the code more readable.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        edge2src = {}
+        for comb in self.g.canonical_etypes:
+            edge2des[comb[1]] = comb[2]
+            edge2src[comb[1]] = comb[0]
```
An implicit assumption is that we cannot have the same edge type across different canonical edge types, which you need to state in the description of the `g` argument.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+
+        ## convert the edge metapath to the node metapath
+        edge2des = {}
+        edge2src = {}
```
It seems that you only need to retrieve `edge2src[self.metapath[0]]` rather than for all edge types in `self.metapath`.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+            nodespath.append(edge2des[edge])
+
+        ##start random walk
+        for idx in tqdm.trange(self.g.number_of_nodes(nodespath[0])):
```
```suggestion
        for idx in tqdm.trange(self.g.num_nodes(nodespath[0])):
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+
+        ##start random walk
+        for idx in tqdm.trange(self.g.number_of_nodes(nodespath[0])):
+            traces = random_walk(g=self.g, nodes=[idx, ] * self.node_repeat, metapath=self.metapath)
```
```suggestion
            traces = random_walk(g=self.g, nodes=[idx] * self.node_repeat, metapath=self.metapath)
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        ##start random walk
+        for idx in tqdm.trange(self.g.number_of_nodes(nodespath[0])):
+            traces = random_walk(g=self.g, nodes=[idx, ] * self.node_repeat, metapath=self.metapath)
+            for tr in traces[0].cpu().numpy():
```
```suggestion
            for tr in traces.cpu().numpy():
```

mufeili replied:
Will different `tr` have different length?

mufeili replied:
We should be able to accelerate this for loop by processing data in batches.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+                if len(line) > 1:
+                    for word in line:
+                        self.token_count += 1
+                        word_frequency[word] = word_frequency.get(word, 0) + 1
```
You can initialize `word_frequency` with `defaultdict(int)`, which initializes any new `word` with `word_frequency[word] = 0`.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+                        self.token_count += 1
+                        word_frequency[word] = word_frequency.get(word, 0) + 1
+
+        # Filter out the number of words less than the minimum frequency
```
```suggestion
        # Filter out the nodes whose number of occurrences is smaller than min_count
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+"""Network Embedding NN modules"""
```
Since this is not for NLP. I think we need to replace all the use of `word`.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+                self.discards[word] = np.sqrt(t / f) + (t / f)
+
+        # get the number of all nodes
+        self.word_count = self.g.number_of_nodes()
```
```suggestion
        self.word_count = self.g.num_nodes()
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        self.word_frequency = word_frequency_filtered
+
+        print("Total embeddings: " + str(self.word_count))
+        print("real embeddings：" + str(len(self.word_frequency)))
```
```suggestion
        print("Real embeddings：" + str(len(self.word_frequency)))
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        """
+        Initialize the embedding parameters
+        """
+        initrange = 1.0 / self.emb_dimension
```
```suggestion
        init_range = 1.0 / self.emb_dimension
```

update the use below accordingly

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        self.u_embeddings = nn.Embedding(self.word_count, self.emb_dimension, sparse=self.sparse)
+        self.v_embeddings = nn.Embedding(self.word_count, self.emb_dimension, sparse=self.sparse)
+
+    def _initTableNegatives(self):
```
```suggestion
    def _init_table_negatives(self):
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        return torch.LongTensor(all_u), torch.LongTensor(all_v), torch.LongTensor(all_neg_v)
+
+    def loader(self, **kwargs):
+        r"""Returns the data loader that contains center node ，positive context node，
```
```suggestion
        r"""Returns the data loader that yields center node ，positive context node，
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+    def loader(self, **kwargs):
+        r"""Returns the data loader that contains center node ，positive context node，
+            and negative samples on the heterogeneous graph random walk.
+        Args:
```
Add an empty line above.

mufeili replied:
```suggestion
        Parameters
        -----------
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        r"""Returns the data loader that contains center node ，positive context node，
+            and negative samples on the heterogeneous graph random walk.
+        Args:
+            **kwargs (optional): Arguments of
```
```suggestion
        **kwargs : dict, optional
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+            **kwargs (optional): Arguments of
+                :class:`torch.utils.data.DataLoader`, such as
+                :obj:`batch_size`, :obj:`shuffle`, :obj:`drop_last` or
+                :obj:`num_workers`.
```
You can reuse the doc of `kwargs` here: https://docs.dgl.ai/generated/dgl.dataloading.DataLoader.html#dgl.dataloading.DataLoader

mufeili replied:
Add a `Returns` section. You can follow the existing DGL APIs.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        pair_catch = []
+        for batch in batches:
+            if len(batch) < 0:
+                continue
```
This never happens, right? If so, you can remove it.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        for batch in batches:
+            if len(batch) < 0:
+                continue
+            word_ids = [w for w in batch if w in self.word_frequency
```
The question is then whether we should directly update `self.walk_dataset` at the beginning.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+                for j, v in enumerate(
+                        word_ids[max(i - self.window_size, 0):i + self.window_size]):
+                    assert u < self.word_count
+                    assert v < self.word_count
```
Can L180 and 181 actually happen? If not, remove these.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+
+        all_u = [u for u, _, _ in pair_catch]
+        all_v = [v for _, v, _ in pair_catch]
+        all_neg_v = [neg_v for _, _, neg_v in pair_catch]
```
You can actually map a list of 3-tuples to three lists at once rather than doing it three times.

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+        self.negpos = (self.negpos + size) % len(self.negatives)
+        if len(response) != size:
+            return np.concatenate((response, self.negatives[0:self.negpos]))
+        return response
```
Here we pre-sample a list of negative node IDs and then sample from them, which is called offline sampling. Is it better to do online sampling or offline sampling here?

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+
+    def forward(self, pos_u, pos_v, neg_v):
+        """
+        Return the loss score given center node,positive context node,and negative samples.
```
```suggestion
        Return the loss score given center nodes, positive context nodes, and negative samples.
```

mufeili commented on python/dgl/nn/pytorch/network_emb.py:
```diff
@@ -0,0 +1,220 @@
+    def forward(self, pos_u, pos_v, neg_v):
+        """
+        Return the loss score given center node,positive context node,and negative samples.
+        """
```
Add `Parameters` and `Returns` section.

Rhett-Ying commented:
Commit ID: 796fc89a4dbdf74a573eebe22091cd2d4f65b3f4

Build ID: 12

Status: ❌ CI test failed in Stage [Torch CPU (Win64) Unit test]. 

Report path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4392/12/12/logs/report.html) 

Full logs path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4392/12/12/logs/cireport.log)

mufeili commented:
@dgl-bot

Rhett-Ying commented:
@dgl-bot

mufeili commented:
@dgl-bot

mufeili commented:
@dgl-bot

mufeili commented:
@dgl-bot

Rhett-Ying commented:
@dgl-bot

mufeili commented:
@dgl-bot

mufeili commented:
@dgl-bot

mufeili commented:
@dgl-bot

mufeili commented:
@dgl-bot

mufeili commented:
Completed via #4660
